### PR TITLE
Expose internal API endpoint for setting and deleting repository meta…

### PR DESCRIPTION
**Pull request by guy-har**

### CODE REVIEW

File: pkg/api/serve.go [hunk 1]
```diff
@@ -62,7 +62,7 @@ func Serve(cfg *config.Config, catalog *catalog.Catalog, middlewareAuthenticator
 	r.Mount("/_health", httputil.ServeHealth())
 	r.Mount("/metrics", promhttp.Handler())
 	r.Mount("/_pprof/", httputil.ServePPROF("/_pprof/"))
-	r.Mount("/swagger.json", http.HandlerFunc(swaggerSpecHandler))
+	r.Mount("/openapi.json", http.HandlerFunc(swaggerSpecHandler))
```
Comment (N-o-Z) [inline pkg/api/serve.go:65]: Is this change related?

**Review by N-o-Z — CHANGES_REQUESTED** [review]
Looks good! One question and one request: Can we please add tests for the Set and Delete (either unit or esti)

File: pkg/api/serve.go (hunk 1 above)
Comment (ghardonag) [reply pkg/api/serve.go:65]: I used [restish](https://github.com/danielgtaylor/restish) to develop locally, I noticed they look for `/openapi.json`, also checked in the [swagger specification](https://swagger.io/specification/)
```
It is RECOMMENDED that the root OpenAPI document be named: openapi.json or openapi.yaml.
```
So, for your question, it's not exactly related, but I believe it's good to add it. 
I can remove it if you prefer

Comment (N-o-Z) [reply pkg/api/serve.go:65]: I don't think it's not a good idea, but the question is should it be replaced or added in addition due to backwards compatibility (@nopcoder )

Comment (nopcoder) [reply pkg/api/serve.go:65]: It is not part of the api/v1, we can document it in the release notes - but I don't see an issue fixing this link to be more aligned with openapi recommendations.

File: pkg/catalog/catalog.go [hunk 2]
```diff
@@ -504,6 +504,37 @@ func (c *Catalog) GetRepositoryMetadata(ctx context.Context, repository string)
 	return c.Store.GetRepositoryMetadata(ctx, repositoryID)
 }
 
+// SetRepositoryMetadata set repository metadata
+func (c *Catalog) UpdateRepositoryMetadata(ctx context.Context, repository string, metadata graveler.RepositoryMetadata) error {
```
Comment (nopcoder) [inline pkg/catalog/catalog.go:508]: - Doc code is not aligned with the func name
- The lower call probably can't check it - if metadata len is 0 just return (relevant also for delete)

File: api/swagger.yml [hunk 3]
```diff
@@ -2589,7 +2610,54 @@ paths:
...
+        Delete repository metadata. Deletes the provided keys from the repository metadata.
+      requestBody:
+        required: true
+        content:
+          application/json:
+            schema:
+              $ref: "#/components/schemas/RepositoryMetadataKeys"
+      responses:
+        204:
+          description: repository metadata deleted successfully
```
Comment (nopcoder) [inline api/swagger.yml:2652]:
```suggestion
          description: repository metadata keys deleted successfully
```

File: api/swagger.yml [hunk 4]
```diff
@@ -2589,7 +2610,54 @@ paths:
...
+          description: too many requests
+        default:
+          $ref: "#/components/responses/ServerError"
+    delete:
+      tags:
+        - internal
+      operationId: deleteRepositoryMetadata
+      summary: delete repository metadata
+      description: |
+        Delete repository metadata. Deletes the provided keys from the repository metadata.
```
Comment (nopcoder) [inline api/swagger.yml:2643]:
```suggestion
        Delete specified keys from the repository's metadata.
```

Also describe if we use 404 in case key to delete not found (I assume not)

File: pkg/permissions/actions.go [hunk 5]
```diff
@@ -16,6 +16,7 @@ var (
 const (
 	ReadRepositoryAction                      = "fs:ReadRepository"
 	CreateRepositoryAction                    = "fs:CreateRepository"
+	UpdateRepositoryAction                    = "fs:UpdateRepository"
```
Comment (nopcoder) [inline pkg/permissions/actions.go:19]: Tricky this one - first I agree with this solution.

My throghts
- Do we want ManageMetadata action?
- Using `Write` vs `Update`?
- Do we want two actions - one for write and one for delete?

(_Note:_ this will require migration)

File: pkg/catalog/catalog.go [hunk 6]
```diff
@@ -504,6 +504,37 @@ func (c *Catalog) GetRepositoryMetadata(ctx context.Context, repository string)
...
+	r, err := c.getRepository(ctx, repository)
+	if err != nil {
+		return err
+	}
+	return c.Store.SetRepositoryMetadata(ctx, r, func(md graveler.RepositoryMetadata) (graveler.RepositoryMetadata, error) {
+		if md == nil {
+			md = make(graveler.RepositoryMetadata)
+		}
```
Comment (nopcoder) [inline pkg/catalog/catalog.go:516]: if `md == nil` just return metadata

**Review by nopcoder — CHANGES_REQUESTED** [review]

File: pkg/permissions/actions.go (hunk 5 above)
Comment (guy-har) [reply pkg/permissions/actions.go:19]: > Do we want ManageMetadata action?
It's a good suggestion, I think update repository is enough 
> Using Write vs Update?
Prefer update because it updates the repository metadata
> Do we want two actions - one for write and one for delete?
IMO leaving update is fine because the action is on the repository level, and even if you choose to change to metadata, it's still on the level of all metadata and not an object (a specific row of metadata)
Don't think it will require a migration for two reasons:
1. It's an internal API 
2. Admins already have FSFullAccess which include `fs:*`

File: api/swagger.yml (hunk 4 above)
Comment (guy-har) [reply api/swagger.yml:2643]: We don't, removing

File: pkg/api/controller_test.go [hunk 7]
```diff
@@ -1188,6 +1188,226 @@ func TestController_DeleteRepositoryHandler(t *testing.T) {
...
+				verifyResponseOK(t, createResp, err)
+
+				resp, err := clt.SetRepositoryMetadataWithResponse(ctx, repoName, apigen.SetRepositoryMetadataJSONRequestBody{Metadata: apigen.RepositoryMetadataSet_Metadata{AdditionalProperties: tt1.properties}})
+				verifyResponseOK(t, resp, err)
+
+				deleteResp, err := clt.DeleteRepositoryMetadataWithResponse(ctx, repoName, apigen.DeleteRepositoryMetadataJSONRequestBody{Keys: tt1.deleteProps})
+				verifyResponseOK(t, deleteResp, err)
+
+				getResp, err := clt.GetRepositoryMetadataWithResponse(ctx, repoName)
+				verifyResponseOK(t, resp, err)
```
Comment (nopcoder) [inline pkg/api/controller_test.go:1360]:
```suggestion
				getResp, err := clt.GetRepositoryMetadataWithResponse(ctx, repoName)
				verifyResponseOK(t, getResp, err)
```

File: pkg/api/controller_test.go [hunk 8]
```diff
@@ -1188,6 +1188,226 @@ func TestController_DeleteRepositoryHandler(t *testing.T) {
...
+				expected:    map[string]string{"foo": "bar"},
+			},
+			{
+				name:        "delete non existing",
+				properties:  map[string]string{"foo": "bar"},
+				deleteProps: []string{"non-existing"},
+				expected:    map[string]string{"foo": "bar"},
+			},
+		}
+		for _, tt1 := range tt {
```
Comment (nopcoder) [inline pkg/api/controller_test.go:1343]: usually tt => tests and tt1 => tt :)

**Review by nopcoder — CHANGES_REQUESTED** [review]
LTGM. small error check missing in the test code.

**Review by nopcoder — APPROVED** [review]
LGTM!

**Review by N-o-Z — APPROVED** [review]
Thanks for adding the tests!
